Packages/ClientRuntime/Sources/Waiters/Matcher.swift

```diff
@@ -0,0 +1,45 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+extension Acceptor {
+
+    /// `Acceptor.Matcher` is a Swift-native equivalent of Smithy matchers:
+    /// https://awslabs.github.io/smithy/2.0/additional-specs/waiters.html#matcher-union
+    public enum Matcher {
```

Comment:
Nit: Feel free to not-change, I leave that to your best judgement but a struct might be more appropriate here rather than an enum. Switching over an enum from within the same enum is usually a smell that the construct might be better modeled as a struct. 

Some rough code to show an example: 
```swift
public struct Matcher {
    private let evaluateMatch: (Input, Result<Output, Error>) -> Bool
    private init(_ evaluateMatch: () -> Bool) {
        self.evaluateMatch = evaluateMatch
    }

    public func hasAMatch(input: Input, result: Result<Output, Error>) -> Bool {
        evaluateMatch(input, result)
    }

    // MARK - Factory

    public static func output(_ f: @escaping (Output) -> Bool) -> Self {
        Matcher { input, result in
            switch result {
            case .succes(let output):
                result f(output)
            case .failure:
                return false
            }
        }
    }

    public static func inputOutput(_ f: @escaping (Input, Output) -> Bool) -> Self {
        // ..
    }

    // ..
}
```

Reply:
I've overhauled Matcher, it's now closer to your suggestion, thanks